spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java

```diff
@@ -379,4 +420,46 @@ public void messageArrived(String topic, MqttMessage mqttMessage) {
 	public void deliveryComplete(IMqttDeliveryToken token) {
 	}
 
+	/**
+	 * Used to complete message arrival when {@link AckMode#MANUAL}.
+	 *
+	 * @since 5.3
+	 */
+	private class AcknowledgmentImpl implements Acknowledgment {
+
+		private final int id;
+
+		private final int qos;
+
+		private final IMqttClient ackClient;
+
+		/**
+		 * Construct an instance with the provided properties.
+		 * @param id the message id.
+		 * @param qos the message QOS.
+		 * @param client the client.
+		 */
+		AcknowledgmentImpl(int id, int qos, IMqttClient client) {
+			this.id = id;
+			this.qos = qos;
+			this.ackClient = client;
+		}
+
+		@Override
+		public void acknowledge() {
+			if (this.ackClient != null || !this.ackClient.equals(MqttPahoMessageDrivenChannelAdapter.this.client)) {
+				try {
+					this.ackClient.messageArrivedComplete(this.id, this.qos);
+				}
+				catch (MqttException e) {
+					throw new IllegalStateException(e);
+				}
+			}
+			else {
+				throw new IllegalStateException("Client has changed");
```

Comment:
I wonder why do we care about this.
We are just have her self-containing `AcknowledgmentImpl`  with the client it has been created.
If that client becomes invalid at the moment of `acknowledge()` we have nothing to do anyway.

What am I missing?

Reply:
How that is possible that `this.ackClient` can be a `null`?
Why do we need this `if..else` logic here at all?

Reply:
It's just for safety. I don't know the client internals and whether `connectionLost` can be called concurrently with `messageArrived` so there could be a race.